File: app/src/main/java/com/crimsonrgames/titanium/htmltester/MainActivity.java
/***

 MainActivity.java

 Copyright 2016 Aldo Pedro Rangel Montiel

 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at

 http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.

 */


package com.crimsonrgames.titanium.htmltester;

import android.content.Intent;
import android.support.design.widget.TabLayout;
import android.support.design.widget.FloatingActionButton;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;

import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentPagerAdapter;
import android.support.v4.view.ViewPager;
import android.os.Bundle;
import android.util.Log;
import android.util.SparseArray;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.webkit.WebView;
import android.widget.EditText;
import android.widget.Toast;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;

public class MainActivity extends AppCompatActivity implements AddTagDialogFragment.OnAddTagDialogFragmentInteractionListener{


    public static final String TAG = "MainActivity";
    public static final String SAVED = "Saved";
    public static final String RESTORED_TO_DEFAULT_TEXT = "Restored to default text";
    public static final String RESTORED_TO_LAST_SAVED_FILE = "Restored to last saved file";
    public static final String TAG_PATTERN = "<%s> </%s>";
    private static String DEFAULT_HTML_STRING = "<html><body><h1>TODO: Write your own code here!!</h1></body></html>";

    private static String DEFAULT_FILE_NAME = "test.html";
    /**
     * The {@link android.support.v4.view.PagerAdapter} that will provide
     * fragments for each of the sections. We use a
     * {@link FragmentPagerAdapter} derivative, which will keep every
     * loaded fragment in memory. If this becomes too memory intensive, it
     * may be best to switch to a
     * {@link android.support.v4.app.FragmentStatePagerAdapter}.
     */
    private SectionsPagerAdapter mSectionsPagerAdapter;

    /**
     * The {@link ViewPager} that will host the section contents.
     */
    private ViewPager mViewPager;
    /**
     * The EditorFragment that will be used to edit the source code
     */
    private EditorFragment mEditorFragment;
    /**
     * The PreviewFragment that will be used to show the WebView with the contents of the source
     * code
     */
    private PreviewFragment mPreviewFragment;

    private String mSourceCode;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);


        mSourceCode = readHTMLFromDefaultFile();
        if(mEditorFragment == null){
            mEditorFragment = EditorFragment.newInstance(mSourceCode);
        }
        if(mPreviewFragment == null){
            mPreviewFragment = PreviewFragment.newInstance(mSourceCode);
        }
        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        // Create the adapter that will return a fragment for each of the three
        // primary sections of the activity.
        mSectionsPagerAdapter = new SectionsPagerAdapter(getSupportFragmentManager());

        // Set up the ViewPager with the sections adapter.
        mViewPager = (ViewPager) findViewById(R.id.container);
        mViewPager.setAdapter(mSectionsPagerAdapter);
        mViewPager.addOnPageChangeListener(new ViewPager.OnPageChangeListener() {
            @Override
            public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {

            }

            @Override
            public void onPageSelected(int position) {
                mEditorFragment = ((EditorFragment)mSectionsPagerAdapter.getRegisteredFragment(0));
                if(mEditorFragment == null) return;
                mSourceCode = mEditorFragment.getmSourceCode();
                mPreviewFragment = ((PreviewFragment)mSectionsPagerAdapter.getRegisteredFragment(1));
                if(mPreviewFragment == null) return;
                mPreviewFragment.setmSourceCode(mSourceCode);

            }

            @Override
            public void onPageScrollStateChanged(int state) {

            }
        });

        TabLayout tabLayout = (TabLayout) findViewById(R.id.tabs);
        tabLayout.setupWithViewPager(mViewPager);

        FloatingActionButton fab = (FloatingActionButton) findViewById(R.id.fab);
        fab.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                mEditorFragment = ((EditorFragment)mSectionsPagerAdapter.getRegisteredFragment(0));
                Intent sendIntent = new Intent();
                sendIntent.setAction(Intent.ACTION_SEND);
                sendIntent.putExtra(Intent.EXTRA_TEXT, mEditorFragment.getmSourceCode());
                sendIntent.setType("text/html");
                startActivity(Intent.createChooser(sendIntent, getResources().getText(R.string.send_to)));
            }
        });


    }


    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        // Inflate the menu; this adds items to the action bar if it is present.
        getMenuInflater().inflate(R.menu.menu_main, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        int id = item.getItemId();
        mEditorFragment = ((EditorFragment)mSectionsPagerAdapter.getRegisteredFragment(0));
        mPreviewFragment = ((PreviewFragment)mSectionsPagerAdapter.getRegisteredFragment(1));
        switch (id){
            case R.id.action_save:
                mSourceCode = mEditorFragment.getmSourceCode();
                this.writeHTMLToDefaultFile(mSourceCode);
                int duration = Toast.LENGTH_SHORT;
                Toast toast = Toast.makeText(this.getApplicationContext(), SAVED, duration);
                toast.show();
                break;
            case R.id.action_restore_default:
                mSourceCode = DEFAULT_HTML_STRING;
                mEditorFragment.setmSourceCode(mSourceCode);
                mPreviewFragment.setmSourceCode(mSourceCode);
                duration = Toast.LENGTH_SHORT;
                toast = Toast.makeText(this.getApplicationContext(), RESTORED_TO_DEFAULT_TEXT, duration);
                toast.show();
                break;
            case R.id.action_return_to_saved:
                mSourceCode = this.readHTMLFromDefaultFile();
                mEditorFragment.setmSourceCode(mSourceCode);
                mPreviewFragment.setmSourceCode(mSourceCode);
                duration = Toast.LENGTH_SHORT;
                toast = Toast.makeText(this.getApplicationContext(), RESTORED_TO_LAST_SAVED_FILE, duration);
                toast.show();
                break;
            case R.id.action_insert_tag:
                showAddTagDialog();
                break;
        }

        return super.onOptionsItemSelected(item);
    }


    private void showAddTagDialog() {
        FragmentManager fm = getSupportFragmentManager();
        AddTagDialogFragment addTagDialog = new AddTagDialogFragment();
        addTagDialog.show(fm,TAG);
    }


    private boolean writeHTMLToDefaultFile(String HTMLString) {

        File file = new File(getExternalFilesDir(null), DEFAULT_FILE_NAME);
        try {
            FileOutputStream stream = new FileOutputStream(file);
            try {
                stream.write(HTMLString.getBytes());
            } catch (IOException e) {
                Log.e(TAG,"Error: "+ e.getLocalizedMessage());

            } finally {
                stream.close();
            }
        } catch (IOException e) {
            Log.e(TAG,"Error: "+ e.getLocalizedMessage());

        }
        return true;
    }

    private String readHTMLFromDefaultFile(){
        String fileContents;
        File file = new File(getExternalFilesDir(null), DEFAULT_FILE_NAME);
        int length = (int) file.length();

        byte[] bytes = new byte[length];
        try{
            FileInputStream in = new FileInputStream(file);
            try {
                in.read(bytes);
                fileContents = new String(bytes);
            } catch (IOException e){
                fileContents = DEFAULT_HTML_STRING;
                Log.e(TAG,"Error: "+ e.getLocalizedMessage());
            }finally {
                in.close();
            }
        }catch (IOException e){
            fileContents = DEFAULT_HTML_STRING;
            Log.e(TAG,"Error: "+ e.getLocalizedMessage());
        }
        Log.d(TAG,fileContents);
        if(fileContents.isEmpty()) fileContents = DEFAULT_HTML_STRING;
        return fileContents;
    }

    /***
     * Overriden from AddTagDialogFragment.OnAddTagDialogFragmentInteractionListener
     * Receives the input from the AddTagDialogFragment
     * @param tag the input tag
     */
    @Override
    public void onFinishTagEditDialog(String tag) {

        mEditorFragment = ((EditorFragment)mSectionsPagerAdapter.getRegisteredFragment(0));
        mEditorFragment.insertTextAtCursorPoint(String.format(TAG_PATTERN, tag,tag));

    }


    /**
     * The Editor Fragment which contains the EditText element that we'll use as the source editor
     */
    public static class EditorFragment extends Fragment {
        /**
         * The EditText element reference
         */
        private EditText mEditor;

        /**
         * Convenience storage of the html code
         */
        private String mSourceCode;

        public EditorFragment() {
        }

        /**
         * Returns a new instance of this fragment with the initializing the mEditor
         * with the initial Code
         * @param initialCode The code which will initialize the mEditor
         */
        public static EditorFragment newInstance(String initialCode) {
            EditorFragment fragment = new EditorFragment();
            fragment.mSourceCode = initialCode;
            return fragment;
        }

        @Override
        public View onCreateView(LayoutInflater inflater, ViewGroup container,
                                 Bundle savedInstanceState) {
            View rootView = inflater.inflate(R.layout.fragment_main, container, false);
            mEditor = (EditText) rootView.findViewById(R.id.editText);
            setTextToEditor();
            return rootView;
        }

        /**
         * Inserts text at the current cursor point of the EditText
         * @param textToAdd the text to add
         * @return if the operation was successful or not
         */
        public boolean insertTextAtCursorPoint(String textToAdd){
            boolean success = false;
            if(mEditor != null ){
                mEditor.getText().insert(mEditor.getSelectionStart(), textToAdd);
                success = true;
            }
            return success;
        }

        public void setTextToEditor(){
            if(mEditor != null ){
                mEditor.setText(mSourceCode);
            }
        }

        //Getters

        public String getmSourceCode() {
            if(mEditor != null ) {
                mSourceCode = mEditor.getText().toString();
            }
            return mSourceCode;

        }

        //Setters

        public void setmSourceCode(String mSourceCode) {
            this.mSourceCode = mSourceCode;
            setTextToEditor();
        }
    }

    /**
     * The Preview Fragment which contains the WebView element used to test the HTML
     */
    public static class PreviewFragment extends Fragment {
        /**
         * The WebView element reference
         */
        private WebView mWebView;

        /**
         * Convenience storage of the html code
         */
        private String mSourceCode;

        public PreviewFragment() {
        }

        /**
         * Returns a new instance of this fragment with the initializing the mEditor
         * with the initial Code
         * @param initialCode The code which will initialize the mEditor
         */
        public static PreviewFragment newInstance(String initialCode) {
            PreviewFragment fragment = new PreviewFragment();
            fragment.mSourceCode = initialCode;
            return fragment;
        }

        @Override
        public View onCreateView(LayoutInflater inflater, ViewGroup container,
                                 Bundle savedInstanceState) {
            View rootView = inflater.inflate(R.layout.fragment_html_preview, container, false);
            mWebView = (WebView) rootView.findViewById(R.id.webView);
            return rootView;
        }
        @Override
        public void setUserVisibleHint(boolean visible)
        {
            super.setUserVisibleHint(visible);
            if (visible && isResumed())
            {
                //Manually call onResume so it loads the HTMLStringAgain
                onResume();
            }
        }

        @Override
        public void onResume()
        {
            super.onResume();
            if (!getUserVisibleHint() || mWebView == null)
            {
                return;
            }
            loadHTMLInWebView();
        }
        private void loadHTMLInWebView(){
            if(mWebView != null){
                mWebView.loadData(mSourceCode, "text/html", null);
            }
        }

        //Getters

        public String getmSourceCode() {
            return mSourceCode;
        }

        //Setters

        public void setmSourceCode(String mSourceCode) {
            this.mSourceCode = mSourceCode;
            loadHTMLInWebView();
        }
    }


    /**
     * A {@link FragmentPagerAdapter} that returns a fragment corresponding to
     * one of the sections/tabs/pages.
     */
    public class SectionsPagerAdapter extends FragmentPagerAdapter {

        SparseArray<Fragment> registeredFragments = new SparseArray<Fragment>();

        public SectionsPagerAdapter(FragmentManager fm) {
            super(fm);
        }

        @Override
        public Fragment getItem(int position) {
            // getItem is called to instantiate the fragment for the given page.
            // Return a EditorFragment (defined as a static inner class below).
            if(mEditorFragment == null){
                mEditorFragment = EditorFragment.newInstance(mSourceCode);
            }
            if(mPreviewFragment == null){
                mPreviewFragment = PreviewFragment.newInstance(mSourceCode);
            }
            return position == 0 ? mEditorFragment : mPreviewFragment;
        }

        @Override
        public Object instantiateItem(ViewGroup container, int position) {
            Fragment fragment = (Fragment) super.instantiateItem(container, position);
            registeredFragments.put(position, fragment);
            return fragment;
        }

        @Override
        public void destroyItem(ViewGroup container, int position, Object object) {
            registeredFragments.remove(position);
            super.destroyItem(container, position, object);
        }

        public Fragment getRegisteredFragment(int position) {
            return registeredFragments.get(position);
        }

        @Override
        public int getCount() {
            // Show 2 total pages.
            return 2;
        }

        @Override
        public CharSequence getPageTitle(int position) {
            switch (position) {
                case 0:
                    return "EDITOR";
                case 1:
                    return "PREVIEW";
            }
            return null;
        }
    }
}

File: app/src/main/java/com/crimsonrgames/titanium/htmltester/MainActivity.kt
/***
 *
 * MainActivity.java
 *
 * Copyright 2016 Aldo Pedro Rangel Montiel
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */


package com.crimsonrgames.titanium.htmltester

import android.content.Intent

import android.os.Bundle
import android.util.Log
import android.util.SparseArray
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.webkit.WebView
import android.widget.EditText
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import androidx.fragment.app.FragmentPagerAdapter
import androidx.viewpager.widget.ViewPager
import com.google.android.material.floatingactionbutton.FloatingActionButton
import com.google.android.material.tabs.TabLayout

import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.io.IOException

class MainActivity : AppCompatActivity(), AddTagDialogFragment.OnAddTagDialogFragmentInteractionListener {
    /**
     * The [android.support.v4.view.PagerAdapter] that will provide
     * fragments for each of the sections. We use a
     * [FragmentPagerAdapter] derivative, which will keep every
     * loaded fragment in memory. If this becomes too memory intensive, it
     * may be best to switch to a
     * [android.support.v4.app.FragmentStatePagerAdapter].
     */
    private var mSectionsPagerAdapter: SectionsPagerAdapter? = null

    /**
     * The [ViewPager] that will host the section contents.
     */
    private var mViewPager: ViewPager? = null
    /**
     * The EditorFragment that will be used to edit the source code
     */
    private var mEditorFragment: EditorFragment? = null
    /**
     * The PreviewFragment that will be used to show the WebView with the contents of the source
     * code
     */
    private var mPreviewFragment: PreviewFragment? = null

    private var mSourceCode: String? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)


        mSourceCode = readHTMLFromDefaultFile()
        if (mEditorFragment == null) {
            mEditorFragment = EditorFragment.newInstance(mSourceCode)
        }
        if (mPreviewFragment == null) {
            mPreviewFragment = PreviewFragment.newInstance(mSourceCode)
        }
        val toolbar = findViewById<View>(R.id.toolbar) as Toolbar
        setSupportActionBar(toolbar)
        // Create the adapter that will return a fragment for each of the three
        // primary sections of the activity.
        mSectionsPagerAdapter = SectionsPagerAdapter(supportFragmentManager)

        // Set up the ViewPager with the sections adapter.
        mViewPager = findViewById<View>(R.id.container) as ViewPager
        mViewPager!!.adapter = mSectionsPagerAdapter
        mViewPager!!.addOnPageChangeListener(object : ViewPager.OnPageChangeListener {
            override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {

            }

            override fun onPageSelected(position: Int) {
                mEditorFragment = mSectionsPagerAdapter!!.getRegisteredFragment(0) as EditorFragment
                if (mEditorFragment == null) return
                mSourceCode = mEditorFragment!!.getmSourceCode()
                mPreviewFragment = mSectionsPagerAdapter!!.getRegisteredFragment(1) as PreviewFragment
                if (mPreviewFragment == null) return
                mPreviewFragment!!.setmSourceCode(mSourceCode)

            }

            override fun onPageScrollStateChanged(state: Int) {

            }
        })

        val tabLayout = findViewById<View>(R.id.tabs) as TabLayout
        tabLayout.setupWithViewPager(mViewPager)

        val fab = findViewById<View>(R.id.fab) as FloatingActionButton
        fab.setOnClickListener {
            mEditorFragment = mSectionsPagerAdapter!!.getRegisteredFragment(0) as EditorFragment
            val sendIntent = Intent()
            sendIntent.action = Intent.ACTION_SEND
            sendIntent.putExtra(Intent.EXTRA_TEXT, mEditorFragment!!.getmSourceCode())
            sendIntent.type = "text/html"
            startActivity(Intent.createChooser(sendIntent, resources.getText(R.string.send_to)))
        }


    }


    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        // Inflate the menu; this adds items to the action bar if it is present.
        menuInflater.inflate(R.menu.menu_main, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        val id = item.itemId
        mEditorFragment = mSectionsPagerAdapter!!.getRegisteredFragment(0) as EditorFragment
        mPreviewFragment = mSectionsPagerAdapter!!.getRegisteredFragment(1) as PreviewFragment
        when (id) {
            R.id.action_save -> {
                mEditorFragment?.let{
                    mSourceCode = it.getmSourceCode()
                    this.writeHTMLToDefaultFile(mSourceCode!!)
                    val duration = Toast.LENGTH_SHORT
                    val toast = Toast.makeText(this.applicationContext, SAVED, duration)
                    toast.show()
                }
            }
            R.id.action_restore_default -> {
                mSourceCode = DEFAULT_HTML_STRING
                mEditorFragment?.let { it.setmSourceCode(mSourceCode)}
                mPreviewFragment?.let{ it.setmSourceCode(mSourceCode)}
                val duration = Toast.LENGTH_SHORT
                val toast = Toast.makeText(this.applicationContext, RESTORED_TO_DEFAULT_TEXT, duration)
                toast.show()
            }
            R.id.action_return_to_saved -> {
                mSourceCode = this.readHTMLFromDefaultFile()
                mEditorFragment!!.setmSourceCode(mSourceCode)
                mPreviewFragment!!.setmSourceCode(mSourceCode)
                val duration = Toast.LENGTH_SHORT
                val toast = Toast.makeText(this.applicationContext, RESTORED_TO_LAST_SAVED_FILE, duration)
                toast.show()
            }
            R.id.action_insert_tag -> showAddTagDialog()
        }

        return super.onOptionsItemSelected(item)
    }


    private fun showAddTagDialog() {
        val input = EditText(this)
        input.setSingleLine()

        val dialog = AlertDialog.Builder(this)
                .setTitle(R.string.title_add_tag)
                .setMessage(R.string.tag_dialog_text)
                .setPositiveButton(R.string.action_insert_Tag){ dialogInterface, i ->
                    mEditorFragment?.let {
                        val tag = input.text.toString()
                        it.insertTextAtCursorPoint(String.format(TAG_PATTERN, tag, tag))
                    }
                }
                .setNegativeButton(R.string.action_cancel){ dialogInterface, i ->
                    dialogInterface.cancel()

                }
                .create()
        dialog.setView(input,20,0,20,0)
        dialog.show()
    }


    private fun writeHTMLToDefaultFile(HTMLString: String): Boolean {

        val file = File(getExternalFilesDir(null), DEFAULT_FILE_NAME)
        try {
            val stream = FileOutputStream(file)
            try {
                stream.write(HTMLString.toByteArray())
            } catch (e: IOException) {
                Log.e(TAG, "Error: " + e.localizedMessage)

            } finally {
                stream.close()
            }
        } catch (e: IOException) {
            Log.e(TAG, "Error: " + e.localizedMessage)

        }

        return true
    }

    private fun readHTMLFromDefaultFile(): String {
        var fileContents: String
        val file = File(getExternalFilesDir(null), DEFAULT_FILE_NAME)
        val length = file.length().toInt()

        val bytes = ByteArray(length)
        try {
            val `in` = FileInputStream(file)
            try {
                `in`.read(bytes)
                fileContents = String(bytes)
            } catch (e: IOException) {
                fileContents = DEFAULT_HTML_STRING
                Log.e(TAG, "Error: " + e.localizedMessage)
            } finally {
                `in`.close()
            }
        } catch (e: IOException) {
            fileContents = DEFAULT_HTML_STRING
            Log.e(TAG, "Error: " + e.localizedMessage)
        }

        Log.d(TAG, fileContents)
        if (fileContents.isEmpty()) fileContents = DEFAULT_HTML_STRING
        return fileContents
    }

    /***
     * Overriden from AddTagDialogFragment.OnAddTagDialogFragmentInteractionListener
     * Receives the input from the AddTagDialogFragment
     * @param tag the input tag
     */
    override fun onFinishTagEditDialog(tag: String) {

        mEditorFragment = mSectionsPagerAdapter!!.getRegisteredFragment(0) as EditorFragment
        mEditorFragment!!.insertTextAtCursorPoint(String.format(TAG_PATTERN, tag, tag))

    }


    /**
     * The Editor Fragment which contains the EditText element that we'll use as the source editor
     */
    class EditorFragment : Fragment() {
        /**
         * The EditText element reference
         */
        private var mEditor: EditText? = null

        /**
         * Convenience storage of the html code
         */
        private var mSourceCode: String? = null

        override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                                  savedInstanceState: Bundle?): View? {
            val rootView = inflater.inflate(R.layout.fragment_main, container, false)
            mEditor = rootView.findViewById<View>(R.id.editText) as EditText
            setTextToEditor()
            return rootView
        }

        /**
         * Inserts text at the current cursor point of the EditText
         * @param textToAdd the text to add
         * @return if the operation was successful or not
         */
        fun insertTextAtCursorPoint(textToAdd: String): Boolean {
            var success = false
            if (mEditor != null) {
                mEditor!!.text.insert(mEditor!!.selectionStart, textToAdd)
                success = true
            }
            return success
        }

        fun setTextToEditor() {
            if (mEditor != null) {
                mEditor!!.setText(mSourceCode)
            }
        }

        //Getters

        fun getmSourceCode(): String {
            if (mEditor != null) {
                return mEditor!!.text.toString()
            }
            return ""

        }

        //Setters

        fun setmSourceCode(mSourceCode: String?) {
            this.mSourceCode = mSourceCode
            setTextToEditor()
        }

        companion object {

            /**
             * Returns a new instance of this fragment with the initializing the mEditor
             * with the initial Code
             * @param initialCode The code which will initialize the mEditor
             */
            fun newInstance(initialCode: String?): EditorFragment {
                val fragment = EditorFragment()
                fragment.mSourceCode = initialCode
                return fragment
            }
        }
    }

    /**
     * The Preview Fragment which contains the WebView element used to test the HTML
     */
    class PreviewFragment : Fragment() {
        /**
         * The WebView element reference
         */
        private var mWebView: WebView? = null

        /**
         * Convenience storage of the html code
         */
        private var mSourceCode: String? = null

        override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                                  savedInstanceState: Bundle?): View? {
            val rootView = inflater.inflate(R.layout.fragment_html_preview, container, false)
            mWebView = rootView.findViewById<View>(R.id.webView) as WebView
            return rootView
        }

        override fun setUserVisibleHint(visible: Boolean) {
            super.setUserVisibleHint(visible)
            if (visible && isResumed) {
                //Manually call onResume so it loads the HTMLStringAgain
                onResume()
            }
        }

        override fun onResume() {
            super.onResume()
            if (!userVisibleHint || mWebView == null) {
                return
            }
            loadHTMLInWebView()
        }

        private fun loadHTMLInWebView() {
            if (mWebView != null) {
                val settings = mWebView!!.settings
                settings.defaultTextEncodingName = "utf-8"
                mWebView!!.loadData(mSourceCode, "text/html; charset=utf-8", "UTF-8")
            }
        }

        //Getters

        fun getmSourceCode(): String? {
            return mSourceCode
        }

        //Setters

        fun setmSourceCode(mSourceCode: String?) {
            this.mSourceCode = mSourceCode
            loadHTMLInWebView()
        }

        companion object {

            /**
             * Returns a new instance of this fragment with the initializing the mEditor
             * with the initial Code
             * @param initialCode The code which will initialize the mEditor
             */
            fun newInstance(initialCode: String?): PreviewFragment {
                val fragment = PreviewFragment()
                fragment.mSourceCode = initialCode
                return fragment
            }
        }
    }


    /**
     * A [FragmentPagerAdapter] that returns a fragment corresponding to
     * one of the sections/tabs/pages.
     */
    inner class SectionsPagerAdapter(fm: FragmentManager) : FragmentPagerAdapter(fm) {

        internal var registeredFragments = SparseArray<Fragment>()

        override fun getItem(position: Int): Fragment {
            // getItem is called to instantiate the fragment for the given page.
            // Return a EditorFragment (defined as a static inner class below).
            if (mEditorFragment == null) {
                mEditorFragment = EditorFragment.newInstance(mSourceCode)
            }
            if (mPreviewFragment == null) {
                mPreviewFragment = PreviewFragment.newInstance(mSourceCode)
            }
            return if (position == 0) mEditorFragment!! else mPreviewFragment!!
        }

        override fun instantiateItem(container: ViewGroup, position: Int): Any {
            val fragment = super.instantiateItem(container, position) as Fragment
            registeredFragments.put(position, fragment)
            return fragment
        }

        override fun destroyItem(container: ViewGroup, position: Int, `object`: Any) {
            registeredFragments.remove(position)
            super.destroyItem(container, position, `object`)
        }

        fun getRegisteredFragment(position: Int): Fragment {
            return registeredFragments.get(position)
        }

        override fun getCount(): Int {
            // Show 2 total pages.
            return 2
        }

        override fun getPageTitle(position: Int): CharSequence? {
            when (position) {
                0 -> return "EDITOR"
                1 -> return "PREVIEW"
            }
            return null
        }
    }

    companion object {


        val TAG = "MainActivity"
        val SAVED = "Saved"
        val RESTORED_TO_DEFAULT_TEXT = "Restored to default text"
        val RESTORED_TO_LAST_SAVED_FILE = "Restored to last saved file"
        val TAG_PATTERN = "<%s> </%s>"
        private val DEFAULT_HTML_STRING = "<html><body><h1>TODO: Write your own code here!!</h1></body></html>"

        private val DEFAULT_FILE_NAME = "test.html"
    }
}
